File: runtime/go/wire_pair_test.go
package runtime

import (
	"context"
	"encoding/json"
	"errors"
	"strings"
	"sync"
	"testing"
	"time"

	"github.com/Bitspark/nightseam/duplex/go"
)

type localTestReturn struct{ send func(duplex.Message) error }

type localTestObserver func(ObserverEvent)

func (observe localTestObserver) Observe(event ObserverEvent) { observe(event) }

func (r *localTestReturn) Send(_ []string, m duplex.Message) error { return r.send(m) }

func testBinding(t *testing.T, endpoint duplex.Endpoint) *Dispatcher {
	t.Helper()
	binding, err := NewDispatcher(endpoint)
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = binding.Close(duplex.CodeNormal, "done") })
	return binding
}

func localPair(t *testing.T, options Options) (duplex.Endpoint, duplex.Endpoint) {
	t.Helper()
	a, b, err := NewWirePair(options)
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() { _ = a.Close(duplex.CodeNormal, "done") })
	return a, b
}

func TestLocalWirePairRoundTripReverseAndIsolation(t *testing.T) {
	a, b := localPair(t, Options{})
	aBinding := testBinding(t, a)
	_, err := HandleWire(aBinding, []string{"reverse"}, func(_ context.Context, raw json.RawMessage) (any, error) { return string(raw), nil })
	if err != nil {
		t.Fatal(err)
	}
	bBinding := testBinding(t, b)
	_, err = HandleWire(bBinding, []string{"call"}, func(ctx context.Context, raw json.RawMessage) (any, error) {
		var result string
		err := CallWire(ctx, b, []string{"reverse"}, raw, &result)
		return result, err
	})
	if err != nil {
		t.Fatal(err)
	}
	var result string
	if err := CallWire(context.Background(), a, []string{"call"}, 7, &result); err != nil || result != "7" {
		t.Fatalf("reverse result %q: %v", result, err)
	}
	x, y := localPair(t, Options{})
	yBinding := testBinding(t, y)
	_, _ = HandleWire(yBinding, []string{"call"}, func(context.Context, json.RawMessage) (any, error) { return "independent", nil })
	_ = a.Close(duplex.CodeNormal, "first pair only")
	if err := CallWire(context.Background(), x, []string{"call"}, nil, &result); err != nil || result != "independent" {
		t.Fatalf("other pair %q: %v", result, err)
	}
}

func TestLocalWirePairRetainsPendingUntilResponse(t *testing.T) {
	a, b := localPair(t, Options{MaxPendingRequests: 1})
	started, release := make(chan struct{}), make(chan struct{})
	bBinding := testBinding(t, b)
	_, _ = HandleWire(bBinding, []string{"hold"}, func(context.Context, json.RawMessage) (any, error) { close(started); <-release; return "done", nil })
	first := make(chan error, 1)
	go func() { var result string; first <- CallWire(context.Background(), a, []string{"hold"}, nil, &result) }()
	<-started
	var result any
	err := CallWire(context.Background(), a, []string{"hold"}, nil, &result)
	var public *PublicError
	if !errors.As(err, &public) || public.Code != "busy" {
		t.Fatalf("pending budget: %v", err)
	}
	close(release)
	if err := <-first; err != nil {
		t.Fatal(err)
	}
	_, _ = HandleWire(bBinding, []string{"next"}, func(context.Context, json.RawMessage) (any, error) { return "reused", nil })
	if err := CallWire(context.Background(), a, []string{"next"}, nil, &result); err != nil {
		t.Fatal(err)
	}
}

func TestLocalWirePairOrderedEventsAndReservedCancel(t *testing.T) {
	a, b := localPair(t, Options{QueueCapacity: 1, MaxPendingRequests: 1})
	bBinding := testBinding(t, b)
	started, cancelled := make(chan struct{}), make(chan struct{})
	_, _ = HandleWire(bBinding, []string{"hold"}, func(ctx context.Context, _ json.RawMessage) (any, error) {
		close(started)
		<-ctx.Done()
		close(cancelled)
		return nil, ctx.Err()
	})
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	answer := make(chan error, 1)
	go func() { answer <- CallWire(ctx, a, []string{"hold"}, nil, nil) }()
	<-started
	entered, release, drained := make(chan struct{}), make(chan struct{}), make(chan struct{})
	var mu sync.Mutex
	var seen []int
	_, _ = bBinding.Register([]string{"event"}, duplex.Receiver{Message: func(_ []string, m duplex.Message) {
		var n int
		_ = json.Unmarshal(m.Frame.Data, &n)
		mu.Lock()
		seen = append(seen, n)
		mu.Unlock()
		if n == 1 {
			close(entered)
			<-release
		} else {
			close(drained)
		}
	}})
	if err := EmitWire(context.Background(), a, []string{"event"}, 1); err != nil {
		t.Fatal(err)
	}
	<-entered
	if err := EmitWire(context.Background(), a, []string{"event"}, 2); err != nil {
		t.Fatal(err)
	}
	cancel()
	if !errors.Is(<-answer, context.Canceled) {
		t.Fatal("caller was not cancelled")
	}
	close(release)
	select {
	case <-cancelled:
	case <-time.After(time.Second):
		t.Fatal("reserved cancel did not arrive")
	}
	<-drained
	mu.Lock()
	defer mu.Unlock()
	if len(seen) != 2 || seen[0] != 1 || seen[1] != 2 {
		t.Fatalf("event order: %v", seen)
	}
}

func TestLocalWirePairOverflowClosesOnlyItsCarrier(t *testing.T) {
	a, b := localPair(t, Options{QueueCapacity: 1})
	bBinding := testBinding(t, b)
	entered, release, ended := make(chan struct{}), make(chan struct{}), make(chan struct{})
	defer close(release)
	_, _ = bBinding.Register([]string{"event"}, duplex.Receiver{Message: func([]string, duplex.Message) { close(entered); <-release }, Closed: func(duplex.Code, string) { close(ended) }})
	if err := EmitWire(context.Background(), a, []string{"event"}, nil); err != nil {
		t.Fatal(err)
	}
	<-entered
	if err := EmitWire(context.Background(), a, []string{"event"}, nil); err != nil {
		t.Fatal(err)
	}
	if err := EmitWire(context.Background(), a, []string{"event"}, nil); !errors.Is(err, ErrBackpressure) {
		t.Fatalf("overflow: %v", err)
	}
	select {
	case <-ended:
	case <-time.After(time.Second):
		t.Fatal("blocked consumer hid closure")
	}
}

func TestLocalWirePairReturnMappingAndFailedResponseRetirement(t *testing.T) {
	a, b := localPair(t, Options{MaxPendingRequests: 1})
	bBinding := testBinding(t, b)
	received := make(chan duplex.Message, 2)
	_, _ = bBinding.Register([]string{"raw"}, duplex.Receiver{Message: func(_ []string, m duplex.Message) { received <- m }})
	failed := errors.New("return failed")
	original := &duplex.ReturnAddress{Wire: &localTestReturn{send: func(duplex.Message) error { return Unpublished(failed) }}}
	if err := a.Send([]string{"raw"}, duplex.Message{Frame: duplex.ProfileFrame{Version: 1, Kind: duplex.ProfileRequest, ID: "c:1", Params: json.RawMessage("null")}, Return: original}); err != nil {
		t.Fatal(err)
	}
	request := <-received
	if request.Return == original {
		t.Fatal("root did not map the return capability")
	}
	if err := a.Send([]string{"raw"}, duplex.Message{Frame: duplex.ProfileFrame{Version: 1, Kind: duplex.ProfileCancel, ID: "c:1"}, Return: original}); err != nil {
		t.Fatal(err)
	}
	if cancelled := <-received; cancelled.Return != request.Return {
		t.Fatal("cancellation used a different return capability")
	}
	err := request.Return.Wire.Send(nil, duplex.Message{Frame: duplex.ProfileFrame{Version: 1, Kind: duplex.ProfileResponse, ID: "c:1", Result: json.RawMessage("null")}})
	if !errors.Is(err, failed) {
		t.Fatalf("return failure lost: %v", err)
	}
	var unpublished *UnpublishedError
	if errors.As(err, &unpublished) {
		t.Fatal("return retained publication proof after dispatch")
	}
	_, _ = HandleWire(bBinding, []string{"next"}, func(context.Context, json.RawMessage) (any, error) { return "reused", nil })
	var result string
	if err := CallWire(context.Background(), a, []string{"next"}, nil, &result); err != nil || result != "reused" {
		t.Fatalf("next: %q, %v", result, err)
	}
}

func TestLocalWirePairPrivateDispatchContext(t *testing.T) {
	a, b := localPair(t, Options{})
	type verifiedKey struct{}
	verified := &struct{ identity string }{"verified locally"}
	dispatch := &wireDispatchContext{ctx: context.WithValue(context.Background(), verifiedKey{}, verified)}
	bBinding := testBinding(t, b)
	_, _ = HandleWire(bBinding, []string{"inspect"}, func(ctx context.Context, _ json.RawMessage) (any, error) {
		if ctx.Value(verifiedKey{}) != verified {
			return nil, errors.New("lost verified dispatch context")
		}
		return "observed", nil
	})
	var result string
	if err := callWire(context.Background(), a, []string{"inspect"}, nil, &result, dispatch); err != nil || result != "observed" {
		t.Fatalf("context: %q, %v", result, err)
	}
}

func TestLocalDispatcherExactAndPrefixRoutes(t *testing.T) {
	a, b := localPair(t, Options{})
	bBinding := testBinding(t, b)
	for _, path := range [][]string{nil, {"a"}} {
		label := "root"
		if len(path) > 0 {
			label = "a"
		}
		_, err := bBinding.RegisterPrefix(path, duplex.Receiver{Message: func(received []string, m duplex.Message) {
			if len(received) == 0 {
				t.Error("callback path lost its origin")
			}
			sendWireResponse(m, json.RawMessage(`"`+label+`"`), nil)
		}})
		if err != nil {
			t.Fatal(err)
		}
	}
	detach, _ := HandleWire(bBinding, []string{"a", "b"}, func(context.Context, json.RawMessage) (any, error) { return "exact", nil })
	var result string
	if err := CallWire(context.Background(), a, []string{"a", "b"}, nil, &result); err != nil || result != "exact" {
		t.Fatalf("exact: %q, %v", result, err)
	}
	detach()
	if err := CallWire(context.Background(), a, []string{"a", "b"}, nil, &result); err != nil || result != "a" {
		t.Fatalf("prefix: %q, %v", result, err)
	}
	if err := CallWire(context.Background(), a, []string{"other"}, nil, &result); err != nil || result != "root" {
		t.Fatalf("root: %q, %v", result, err)
	}
}

func TestLocalWirePairDeadlineRetainsNoncooperativeHandlerBudget(t *testing.T) {
	a, b := localPair(t, Options{RequestTimeout: 15 * time.Millisecond, MaxConcurrentHandlers: 1})
	started, cancelled, release := make(chan struct{}), make(chan struct{}), make(chan struct{})
	defer close(release)
	bBinding := testBinding(t, b)
	_, _ = HandleWire(bBinding, []string{"hold"}, func(ctx context.Context, _ json.RawMessage) (any, error) {
		close(started)
		<-ctx.Done()
		close(cancelled)
		<-release
		return nil, nil
	})
	first := make(chan error, 1)
	go func() { first <- CallWire(context.Background(), a, []string{"hold"}, nil, nil) }()
	<-started
	var public *PublicError
	if err := <-first; !errors.As(err, &public) || public.Code != "cancelled" {
		t.Fatalf("deadline: %v", err)
	}
	select {
	case <-cancelled:
	case <-time.After(time.Second):
		t.Fatal("deadline did not cancel handler")
	}
	err := CallWire(context.Background(), a, []string{"hold"}, nil, nil)
	if !errors.As(err, &public) || public.Code != "busy" {
		t.Fatalf("handler budget: %v", err)
	}
}

func TestLocalWirePairStalledEventDeadlineIsObserved(t *testing.T) {
	pressure := make(chan Backpressure, 1)
	a, b := localPair(t, Options{WriteTimeout: 15 * time.Millisecond, Observer: localTestObserver(func(event ObserverEvent) {
		if event, ok := event.(Backpressure); ok && event.Stalled {
			pressure <- event
		}
	})})
	bBinding := testBinding(t, b)
	release, closed := make(chan struct{}), make(chan struct{})
	defer close(release)
	_, _ = bBinding.Register([]string{"event"}, duplex.Receiver{Message: func([]string, duplex.Message) { <-release }, Closed: func(duplex.Code, string) { close(closed) }})
	if err := EmitWire(context.Background(), a, []string{"event"}, nil); err != nil {
		t.Fatal(err)
	}
	select {
	case event := <-pressure:
		if event.Deadline != 15*time.Millisecond {
			t.Fatalf("deadline: %v", event.Deadline)
		}
	case <-time.After(time.Second):
		t.Fatal("stalled event was not observed")
	}
	<-closed
	if err := EmitWire(context.Background(), a, []string{"event"}, nil); !errors.Is(err, ErrClosed) {
		t.Fatalf("closed pair: %v", err)
	}
}

func TestLocalWirePairOversizedResponseUsesBoundedFallback(t *testing.T) {
	a, b := localPair(t, Options{MaxFrameBytes: 512})
	bBinding := testBinding(t, b)
	_, _ = HandleWire(bBinding, []string{"large"}, func(context.Context, json.RawMessage) (any, error) { return strings.Repeat("x", 2048), nil })
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()
	var result string
	err := CallWire(ctx, a, []string{"large"}, nil, &result)
	var public *PublicError
	if !errors.As(err, &public) || public.Code != "internal" {
		t.Fatalf("oversized response: %v", err)
	}
}

File: runtime/ts/src/wire-pair.test.ts
import assert from 'node:assert/strict';
import test from 'node:test';
import { wirePair } from './wire-pair.ts';
import { callWire, handleWire, emitWire, onWireEvent, setWireContext } from './wire.ts';
import type { Message, ReturnAddress, Wire } from '@bitspark/bitwire';
import { createDispatcher } from './dispatcher.ts';

function deferred<T = void>() {
  let resolve!: (value: T) => void;
  const promise = new Promise<T>((yes) => {
    resolve = yes;
  });
  return { promise, resolve };
}

test('local pair supports reverse calls and independent construction', async (t) => {
  const [a, b] = wirePair();
  const [x, y] = wirePair();
  const leftDispatcher = createDispatcher(a);
  const rightDispatcher = createDispatcher(b);
  const independentDispatcher = createDispatcher(y);
  t.after(() => {
    leftDispatcher.close();
    rightDispatcher.close();
    independentDispatcher.close();
    a.close();
    x.close();
  });
  handleWire(leftDispatcher, ['reverse'], (value) => value);
  handleWire(rightDispatcher, ['call'], (value) => callWire(b, ['reverse'], value));
  handleWire(independentDispatcher, ['call'], () => 'independent');
  assert.equal(await callWire(a, ['call'], 7), 7);
  a.close();
  assert.equal(await callWire(x, ['call']), 'independent');
});

test('local pending budget lives until the response and can then be reused', async (t) => {
  const [a, b] = wirePair({ maxPendingRequests: 1 });
  const dispatcher = createDispatcher(b);
  t.after(() => {
    dispatcher.close();
    a.close();
  });
  const started = deferred(),
    release = deferred();
  handleWire(dispatcher, ['hold'], async () => {
    started.resolve();
    await release.promise;
    return 1;
  });
  const first = callWire(a, ['hold']);
  await started.promise;
  await assert.rejects(callWire(a, ['hold']), { code: 'busy' });
  release.resolve();
  assert.equal(await first, 1);
  assert.equal(await callWire(a, ['hold']), 1);
});

test('local events await serially and cancellation owns capacity outside the full data queue', async (t) => {
  const [a, b] = wirePair({ queueCapacity: 1, maxPendingRequests: 1 });
  const dispatcher = createDispatcher(b);
  t.after(() => {
    dispatcher.close();
    a.close();
  });
  const started = deferred(),
    cancelled = deferred(),
    entered = deferred(),
    release = deferred(),
    drained = deferred();
  const seen: number[] = [];
  handleWire(dispatcher, ['hold'], async (_params, context) => {
    started.resolve();
    await new Promise<void>((resolve) =>
      context.signal.addEventListener(
        'abort',
        () => {
          cancelled.resolve();
          resolve();
        },
        { once: true },
      ),
    );
  });
  onWireEvent(dispatcher, ['event'], async (value) => {
    seen.push(value as number);
    if (value === 1) {
      entered.resolve();
      await release.promise;
    } else drained.resolve();
  });
  const controller = new AbortController();
  const first = callWire(a, ['hold'], null, { signal: controller.signal }).catch((error: unknown) => error);
  await started.promise;
  emitWire(a, ['event'], 1);
  await entered.promise;
  emitWire(a, ['event'], 2);
  controller.abort();
  assert.equal(((await first) as { code: string }).code, 'cancelled');
  release.resolve();
  await Promise.all([cancelled.promise, drained.promise]);
  assert.deepEqual(seen, [1, 2]);
});

test('a full local carrier closes even while its event consumer is held', async (t) => {
  const [a, b] = wirePair({ queueCapacity: 1 });
  t.after(() => a.close());
  const entered = deferred(),
    release = deferred(),
    ended = deferred();
  b.receive({
    message: async () => {
      entered.resolve();
      await release.promise;
    },
    closed: () => ended.resolve(),
  });
  emitWire(a, ['event']);
  await entered.promise;
  emitWire(a, ['event']);
  assert.throws(() => emitWire(a, ['event']), { code: 'busy' });
  await ended.promise;
  release.resolve();
});

test('request and cancellation use one mapped return capability and responses retire on throwing returns', async (t) => {
  const [a, b] = wirePair({ maxPendingRequests: 1 });
  const dispatcher = createDispatcher(b);
  t.after(() => {
    dispatcher.close();
    a.close();
  });
  const received = deferred<Message>(),
    cancelled = deferred<Message>();
  dispatcher.register(['raw'], {
    message: (_path, message) => {
      (message.frame.kind === 'request' ? received : cancelled).resolve(message);
    },
  });
  const returning: ReturnAddress = {
    wire: {
      send: () => {
        throw new Error('return failed');
      },
    },
  };
  a.send(['raw'], { frame: { version: 1, kind: 'request', id: 'c:1', params: null }, return: returning });
  const request = await received.promise;
  assert.notEqual(request.return, returning);
  a.send(['raw'], { frame: { version: 1, kind: 'cancel', id: 'c:1' }, return: returning });
  assert.equal((await cancelled.promise).return, request.return);
  assert.throws(
    () => request.return!.wire.send([], { frame: { version: 1, kind: 'response', id: 'c:1', result: null } }),
    /return failed/,
  );
  handleWire(dispatcher, ['next'], () => 'reused');
  assert.equal(await callWire(a, ['next']), 'reused');
});

test('local roots preserve private dispatch context and keep metadata explicit', async (t) => {
  const [a, b] = wirePair();
  const dispatcher = createDispatcher(b);
  t.after(() => {
    dispatcher.close();
    a.close();
  });
  const verified = Object.freeze({ identity: 'verified locally' });
  const source = { wire: a, signal: new AbortController().signal, requestId: 'physical:17' };
  Object.defineProperty(source, 'verified', { value: verified });
  const traced: Wire = {
    ...a,
    send: (path, message) => {
      if (message.frame.kind === 'request')
        setWireContext(message.return!, { context: source, panic: () => {}, maxFrameBytes: 1024 });
      a.send(path, message);
    },
  };
  handleWire(dispatcher, ['inspect'], (_value, context) => {
    assert.equal((context as typeof context & { verified: unknown }).verified, verified);
    assert.equal(context.requestId, 'physical:17');
    assert.deepEqual(context.meta, { explicit: 'yes' });
    return 'observed';
  });
  assert.equal(await callWire(traced, ['inspect'], null, { meta: { explicit: 'yes' } }), 'observed');
});

test('local exact routes win over the longest namespace and detach reveals fallback', async (t) => {
  const [a, b] = wirePair();
  const dispatcher = createDispatcher(b);
  t.after(() => {
    dispatcher.close();
    a.close();
  });
  const seen: string[] = [];
  const receiver = (label: string) => ({
    message: (path: readonly string[], message: Message) => {
      seen.push(`${label}:${path.join('/')}`);
      message.return!.wire.send([], {
        frame: { version: 1, kind: 'response', id: (message.frame as { id: string }).id, result: label },
      });
    },
  });
  dispatcher.registerPrefix([], receiver('root'));
  dispatcher.registerPrefix(['a'], receiver('a'));
  const detach = handleWire(dispatcher, ['a', 'b'], () => 'exact');
  assert.equal(await callWire(a, ['a', 'b']), 'exact');
  detach();
  assert.equal(await callWire(a, ['a', 'b']), 'a');
  assert.equal(await callWire(a, ['other']), 'root');
  assert.deepEqual(seen, ['a:a/b', 'root:other']);
});

test('local deadline cancels handlers but retains noncooperative handler capacity', async (t) => {
  const [a, b] = wirePair({ requestTimeoutMs: 15, maxConcurrentHandlers: 1 });
  const dispatcher = createDispatcher(b);
  t.after(() => {
    dispatcher.close();
    a.close();
  });
  const release = deferred(),
    started = deferred(),
    cancelled = deferred();
  let invocations = 0;
  handleWire(dispatcher, ['hold'], async (_value, context) => {
    invocations++;
    started.resolve();
    context.signal.addEventListener('abort', () => cancelled.resolve(), { once: true });
    await release.promise;
  });
  const first = callWire(a, ['hold']).catch((error: unknown) => error);
  await started.promise;
  assert.equal(((await first) as { code: string }).code, 'cancelled');
  await cancelled.promise;
  await assert.rejects(callWire(a, ['hold']), { code: 'busy' });
  assert.equal(invocations, 1);
  release.resolve();
});

test('local queued payload snapshots cannot be changed by the sender', async (t) => {
  const [a, b] = wirePair();
  const dispatcher = createDispatcher(b);
  t.after(() => {
    dispatcher.close();
    a.close();
  });
  const seen = deferred<unknown>();
  onWireEvent(dispatcher, ['event'], (value) => seen.resolve(value));
  const data = { nested: ['original'] };
  a.send(['event'], { frame: { version: 1, kind: 'event', data } });
  data.nested[0] = 'mutated';
  assert.deepEqual(await seen.promise, { nested: ['original'] });
});

test('a stalled local event reaches the configured deadline and observer', async (t) => {
  const pressure = deferred<number>();
  const [a, b] = wirePair({
    writeTimeoutMs: 15,
    observer: {
      observe: (event) => {
        if (event.type === 'backpressure' && event.stalled) pressure.resolve(event.deadlineMs);
      },
    },
  });
  t.after(() => a.close());
  const release = deferred(),
    closed = deferred();
  b.receive({ message: () => release.promise, closed: () => closed.resolve() });
  emitWire(a, ['event']);
  assert.equal(await pressure.promise, 15);
  await closed.promise;
  assert.throws(() => emitWire(a, ['event']), { code: 'disconnected' });
  release.resolve();
});

test('an oversized local response settles through the bounded internal fallback', async (t) => {
  const [a, b] = wirePair({ maxFrameBytes: 512 });
  const dispatcher = createDispatcher(b);
  t.after(() => {
    dispatcher.close();
    a.close();
  });
  handleWire(dispatcher, ['large'], () => 'x'.repeat(2048));
  await assert.rejects(callWire(a, ['large'], null, { timeoutMs: 1000 }), { code: 'internal' });
});
